### PR TITLE
Move Notification Logic from Discussion Store to Firebase Functions

### LINKED ISSUE #3563: Move Notification Logic from Discussion Store to Firebase Functions
Refactor the notification logic currently housed within the discussion store and move it to Firebase Functions for better scalability, maintainability, and separation of concerns.

Steps: 
- [ ] Remove from discussion store
- [ ] Add to `functions/src/discussionUpdates/index.ts` once its merged from #3557 
- [ ] Test

### PR DESCRIPTION
**Pull request by goratt12**
…ebase functions

PR Checklist

- [ ] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)

PR Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Developer experience (improves developer workflows for contributing to the project)

## Description

Move Notification Logic from Discussion Store to Firebase Functions
Added tests

## Git Issues

Closes #3563 

## Screenshots/Videos

_If useful, provide screenshot or capture to highlight main changes_

---

## What happens next?

Thanks for the contribution! We try to make sure all PRs are reviewed ahead of our monthly maintainers call (first Monday of the month)

If the PR is working as intended it'll be merged and included in the next platform release, if not changes will be requested and re-reviewed once updated.

If you need more immediate feedback you can try reaching out on Discord in the [Community Platform `development` channel](https://discord.com/channels/586676777334865928/938781727017558018).

### CODE REVIEW

**Comment by goratt12** [comment]
Tested manually and it's working.
tested on both research and questions - nested and regular comments.

File: functions/src/discussionUpdates/index.test.ts [hunk 1]
```diff
@@ -128,5 +177,299 @@ describe('discussionUpdates', () => {
...
+          .get()
+      ).docs[0].data() as IUserDB
+
+      expect(researchCollaborator.notifications).toHaveLength(1)
+      expect(researchCollaborator.notifications[0].relevantUrl).toContain(
+        'fake-research',
+      )
+      expect(researchCollaborator.notifications[0].relevantUrl).toContain(
+        commentId,
+      )
```
Comment (benfurber) [inline functions/src/discussionUpdates/index.test.ts:255]: As your code is generating a whole notification object, I think you should be expecting a whole notification object e.g. `expect(researchCollaborator.notifications[0]).toEqual(notification)`

**Review by benfurber — COMMENTED** [review]
Nice one. Just leaving some code level comments initially. Especially as the cypress tests are failing (which I think we should expect as the functions the ci env are running won't include these changes) I'll checkout the branch and test myself.

File: functions/src/discussionUpdates/index.test.ts [hunk 2]
```diff
@@ -128,5 +177,299 @@ describe('discussionUpdates', () => {
...
+    })
+
+    it('create nested comment and notify parent comment creator - research', async () => {
```
[the remark this answers is not in the document]
Comment (goratt12) [reply functions/src/discussionUpdates/index.test.ts:258]: Should I change the title from 'create nested comment and notify parent comment creator - research' to 'create nested comment and notify parent comment creator'? without the ' - research'

**Comment by benfurber** [comment]
I've deployed these functions to the ci env. Will rerun the tests now and hopefully all will pass...

**Comment by benfurber** [comment]
@goratt12. Tests still failing unfortunately. Can I leave it to you please to confirm the functions deployed as expected, etc. and make changes accordingly?

**Comment by goratt12** [comment]
> @goratt12. Tests still failing unfortunately. Can I leave it to you please to confirm the functions deployed as expected, etc. and make changes accordingly?

Could you please give me permissions to the onearmy-test-ci Firebase project?

**Comment by benfurber** [comment]
@goratt12 The test comment issue, can you give more of the detail we talked about yesterday please. 

Looks like you've still got a merge conflict? Happy to jump on a final manual test once that's sorted.

File: functions/src/discussionUpdates/index.ts [hunk 3]
```diff
@@ -1,21 +1,35 @@
...
  *********************************************************************/
 
 export const handleDiscussionUpdate = functions
   .runWith({ memory: '512MB' })
   .firestore.document(`${DB_ENDPOINTS.discussions}/{id}`)
-  .onUpdate(async (change, context) => {
-    await updateDocument(change)
+  .onUpdate((change, context) => {
+    try {
+      return updateDocument(change)
```
Comment (mariojsnunes) [inline functions/src/discussionUpdates/index.ts:29]: this needs to be awaited, otherwise won't enter the catch block in case of error

Comment (goratt12) [reply functions/src/discussionUpdates/index.ts]: Correct, fixed it

**Comment by mariojsnunes** [comment]
For the tests, if they are now invoking cloud functions, maybe we could mock the request?
Did that recently here:
https://github.com/mariojsnunes/community-platform/blob/4f0e648e56a14a9560362879561798c22551bb9e/packages/cypress/src/integration/profile.spec.ts#L18-L21

**Comment by goratt12** [comment]
> For the tests, if they are now invoking cloud functions, maybe we could mock the request? Did that recently here: https://github.com/mariojsnunes/community-platform/blob/4f0e648e56a14a9560362879561798c22551bb9e/packages/cypress/src/integration/profile.spec.ts#L18-L21

That's a good idea, but the functions are internally triggered by Firestore changes, so we cannot intercept them there. The problem is that the function handleDiscussionUpdate listens for changes at this path: discussions_rev20231022/{id}. In our E2E tests, the Firestore collection is {run_id}_discussions_rev20231022. We considered changing the function to listen for changes with a wildcard path, but we decided that altering the environment for the tests is not the correct approach.

**Comment by benfurber** [comment]
@goratt12 We decided what to do about this didn't we? If we can't test it without big setup changes, we can't test it without big setup changes.

**Comment by goratt12** [comment]
> @goratt12 We decided what to do about this didn't we? If we can't test it without big setup changes, we can't test it without big setup changes.

Agreed, so currently I'm leaving the relevant tests commented with the explanation

**Pull request closed without merging**